scrapy_poet/_request_fingerprinter.py

```diff
@@ -0,0 +1,171 @@
+                "been configured in the DOWNLOADER_MIDDLEWARES setting?"
+            )
+
+        def _get_deps(self, request: Request) -> Optional[List[str]]:
+            """Return a JSON-serializable structure that uniquely identifies the
+            dependencies requested by the request, or None if dependency injection
+            is not required."""
+            plan = self._injector.build_plan(request)
+            root_deps = plan[-1][1]
+            if not root_deps:
+                return None
+            return sorted(
+                [
+                    _serialize_dep(cls)
+                    for cls in root_deps.values()
+                    if cls not in self.IGNORED_UNANNOTATED_DEPS
+                ]
+            )
+
+        def get_deps_key(self, request: Request) -> Optional[bytes]:
+            """Return a JSON array as bytes that uniquely identifies the
+            dependencies requested through scrapy-poet injection that could
+            impact the request, or None if there are no such dependencies."""
+            callback = get_callback(request, self._crawler.spider)
+            if callback in self._callback_cache:
+                return self._callback_cache[callback]
+
+            deps = self._get_deps(request)
+            if not deps:
+                return None
+
+            deps_key = json.dumps(deps, sort_keys=True).encode()
+            self._callback_cache[callback] = deps_key
+            return self._callback_cache[callback]
+
+        def serialize_page_params(self, request: Request) -> Optional[bytes]:
+            """Returns a JSON object as bytes that represents the page params,
+            or None if there are no page params or they are not
+            JSON-serializable."""
+            page_params = request.meta.get("page_params", None)
+            if not page_params:
+                return None
+
+            try:
+                return json.dumps(page_params, sort_keys=True).encode()
+            except TypeError:
+                if not self._saw_unserializable_page_params:
+                    self._saw_unserializable_page_params = True
+                    logger.warning(
+                        f"Cannot serialize page params {page_params!r} of "
+                        f"request {request} as JSON. This can be an issue if "
+                        f"you have requests that are identical except for "
+                        f"their page params, because unserializable page "
+                        f"params are treated the same as missing or empty "
+                        f"page params for purposes of request fingerprinting "
+                        f"(see "
+                        f"https://docs.scrapy.org/en/latest/topics/request-response.html#request-fingerprints). "
+                        f"This will be the only warning about this issue, "
+                        f"other requests might be also affected."
+                    )
+                return None
```

Comment:
Given [we expect JSON-serializable page params for caching](https://github.com/scrapinghub/web-poet/blob/d45aa1e74aab15466100cbf71b1e655d6c6572f0/web_poet/serialization/functions.py#L175), I went with expecting so, and logging a warning and fingerprinting as if there were no params when that’s not the case.

I thought about keeping a fixed size cache for page params, so save json.dumps calls on scenarios where the same page params are often used, but I was unsure about it.

I also thought about using a random value instead of treating them the same as missing or empty params, to force a different fingerprint, but that could cause a different fingerprint for requests with the same (unserializable) params, which I think is worse (can destroy more data through duplicate request dropping).